src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/descriptor_set_data.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2020 The gf-layers Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VKLAYER_GF_AMBER_SCOOP_DESCRIPTOR_SET_DATA_H
+#define VKLAYER_GF_AMBER_SCOOP_DESCRIPTOR_SET_DATA_H
+
+#include <vulkan/vulkan.h>
+
+#include <cstdint>
+#include <unordered_map>
+#include <vector>
+
+#include "VkLayer_GF_amber_scoop/create_info_wrapper.h"
+
+namespace gf_layers::amber_scoop_layer {
+
+// This class is used to store descriptor set's state.
+class DescriptorSetData {
+ public:
+  // Type alias used to clarify the maps' key types.
+  using BindingNumber = uint32_t;
+
+  // Flattened descriptor types. Used f.ex. to easily choose where the
+  // descriptor writes are taken from and where they are stored.
+  enum FlattenedDescriptorType {
+    IMAGE_SAMPLER,
+    BUFFER,
+    TEXEL_BUFFER,
+    UNSUPPORTED
+  };
+
+  // Constructor. Stores the given pointer to the layout used to allocate this
+  // descriptor set and allocates the vectors where the information about
+  // descriptors will be stored when updating the descriptor set.
+  explicit DescriptorSetData(const DescriptorSetLayoutData* layout_data);
+
+  // Get the descriptor bindings of buffer type.
+  [[nodiscard]] const std::unordered_map<BindingNumber,
+                                         std::vector<VkDescriptorBufferInfo>>*
+  GetDescriptorBufferBindings() const {
+    return &descriptor_buffer_bindings_;
+  }
+
+  // Get pointer to the layout used to create this descriptor set.
+  [[nodiscard]] const DescriptorSetLayoutData* GetDescriptorSetLayoutData()
+      const {
+    return descriptor_set_layout_data_;
+  }
+
+  // Updates the descriptor sets' state with the given VkWriteDescriptorSet.
+  void WriteDescriptorSet(const VkWriteDescriptorSet& write_descriptor_set);
+
+  // TODO(ilkkasaa): implement CopyDescriptorSet function.
+  /*
+  void CopyDescriptorSet(const VkCopyDescriptorSet& copy_descriptor_set) {
+    RUNTIME_ASSERT_MSG(false, "CopyDescriptorSet not implemented.");
+  }*/
+
+ private:
+  // Layout used to allocate this descriptor set.
+  const DescriptorSetLayoutData* descriptor_set_layout_data_;
```

Comment:
So it looks like `DescriptorSetData` does not own the `DescriptorSetLayoutData`. I think this is almost OK because the Vulkan specs says you must not update a descriptor set after destroying the layout. However, you also have a getter that is used when processing a draw command; I believe the layout can be destroyed by this point. Thus, I think we will have to make a deep copy of the layout data and `DescriptorSetData` will own that copy. I could be wrong though.

Reply:
Good catch! I need to do some investigation if the layout can be destroyed when the draw command is being processed.

Reply:
Yeah, you're right! The layout can be destroyed by that point. I'll change it to make a copy of the `DescriptorSetData`. I guess we could also use shared ownership here, but I think the performance gain achieved by avoiding deep copy of one class and a struct isn't good enough reason to use shared_ptr.